Upstream shape api part 4 (of 4)

Elv13 opened the pull request:
Ok, here is  giant PR I once said I would not submit before a release. But here it is anyway. I first said it existed back in January, so it should not be such a big surprise. I cleaned and completed the various parts since May.

So first, it does complete the shape API

-- \* `client`s (see `gears.surface.apply_shape_bounding`)
 -- \* `wibox`es (see `wibox.shape`)
 -- \* All widgets (see `wibox.container.background`)
 -- \* The progressbar (see `wibox.widget.progressbar.bar_shape`)
 -- \* The graph (see `wibox.widget.graph.step_shape`)
 -- \* The checkboxes (see `wibox.widget.checkbox.check_shape`)
 -- \* Images (see `wibox.widget.imagebox.clip_shape`)
 -- \* Sliders (see `wibox.widget.slider.handle_shape`)
 -- \* The taglist tags (see `awful.widget.taglist`)
 -- \* The tasklist clients (see `awful.widget.tasklist`)
 -- \* The tooltips (see `awful.tooltip`)

It also adds a lot of doc. Actually, 50% of the PR is doc related. It adds tens of examples, dozens of images and 2 whole new documentation pages. As @psychon asked, I didn't try to implement the new auto-ganarated pages using CMake. I used lua scripts instead. So now all beautiful variables are documented and an index is auto-generated. The other pages is to documented `rc.lua`. It could help meta-document any user file too (in case the config is eventually splitted). Beside this, I manually read the whole ldoc documentation and fixed the issues I saw.

And yeah, this one is **HUGE**. Sorry about that. I could make some iterations. But why hiding the whole picture.

This commit also add the grid layout, just because I had it ready (it was in Radical since a while). I know luacheck still find issues and that there is a few obvious bugs, but as far as development goes, this is "done".

![slider1](https://cloud.githubusercontent.com/assets/340384/17689999/a8118f6c-6359-11e6-945a-0510b55d65db.png)

Edit: And thanks to @actionless for testing this a little and contributing the notification part.

Elv13 commented:
So, just to confirm what I said. Here is the breakdown:

1627 lines of doc
1304 lines of tests / examples
563 lines of meta-documentation / linters
996 empty lines
925 lines of new code

So it is actually not _that_ big.

psychon commented on lib/wibox/widget/progressbar.lua:
```diff
@@ -181,16 +182,24 @@ function progressbar:set_value(value)
 -- @param height The height to set.
-function progressbar:set_height(height)
+-- @deprecated set_height
+function progressbar:set_height(height) --luacheck: no unused_args
```
I don't think that this luacheck comment is needed, is it? (Same comment elsewhere here)

(I tried looking at the "complete diff" for this for a quick check, but that's just too big and mixes lots of things. Thus, I'll take a quick look through the individual commits).

psychon commented on lib/wibox/widget/progressbar.lua:
```diff
@@ -144,10 +148,31 @@ function progressbar:set_value(value)
+    -- Restore the value that was trimmed due to a race condition
+    if self._private.value_cache and self._private.value_cache <= max_value then
+        self._private.value = self._private.value_cache
+    end
```
Hm... How about having `set_value` only do `self._private.value = value` and move the `math.min(max_value, math.max(0, value))` into `:draw()`? I.e. we never trim the value that is saved, but only apply the triming while drawing the widget. That might be simpler to understand than this version.

psychon commented on lib/wibox/container/radialprogressbar.lua:
```diff
@@ -0,0 +1,264 @@
+    self:emit_signal("widget::layout_changed")
+end
+
+--- Get the number of children element
```
`Get the number of children element` doesn't sound like it describes what this function does, does it?

psychon commented on lib/wibox/container/radialprogressbar.lua:
```diff
@@ -0,0 +1,264 @@
+    self:emit_signal("widget::layout_changed")
+end
+
+--- Reset this layout. The widget will be removed and the rotation reset.
```
Which rotation is being reset?

psychon commented on lib/wibox/container/radialprogressbar.lua:
```diff
@@ -0,0 +1,264 @@
+    radialprogressbar["set_"..v.."_padding"] = function(self, val)
+        self._private.paddings = self._private.paddings or {}
+        self._private.paddings[v] = val
+        self:emit_signal("widget::redraw_needed")
```
Shouldn't it (also) be `layout_changed`? The padding is used by `:fit` and `:layout` (and `:before_draw_children`).

psychon commented on lib/wibox/container/radialprogressbar.lua:
```diff
@@ -0,0 +1,264 @@
+    if val > self._private.max_value then
+        self:set_max_value(val)
+    elseif val < self._private.min_value then
+        self:set_min_value(val)
```
Does this work? As far as I can tell, there is no `max/min`, but `maximum/minimum`.

psychon commented on lib/wibox/container/radialprogressbar.lua:
```diff
@@ -0,0 +1,264 @@
+    radialprogressbar["set_"..prop] = function(self, value)
+        self._private[prop] = value
+        self:emit_signal("property::"..prop)
+        self:emit_signal("widget::redraw_needed")
```
`"paddings"` also influences the layout and thus needs `widget::layout_changed`.

psychon commented on lib/wibox/container/radialprogressbar.lua:
```diff
@@ -0,0 +1,264 @@
+end
+
+--- Returns a new radialprogressbar layout. A radialprogressbar layout radialprogressbars a given widget. Use
+-- :set_widget() to set the widget and :set_direction() for the direction.
```
There is no `set_direction`, is there?

psychon commented on lib/wibox/widget/checkbox.lua:
```diff
@@ -0,0 +1,253 @@
+
+    util.table.crush(ret, checkbox)
+
+    ret._private.style = args.style or beautiful.checkbox_style or "default"
```
This `style` value is never used

psychon commented on lib/wibox/widget/checkbox.lua:
```diff
@@ -0,0 +1,253 @@
+
+return setmetatable({}, { __call = function(_, ...) return new(...) end})
+
+-- kate: space-indent on; indent-width 4; replace-tabs on;
```
Heh, someone is working against vim modelines :-)

blueyed replied:
I would argue to not use any modelines, but if so, only Vim ones - which probably can be parsed by other editors, too?!
FWIW, we have an `.editorconfig` already.

psychon replied:
Removing vim modelines would also be fine with me if many people are against them. We only have them because JD added them "Back Then" (tm).

psychon commented on lib/wibox/widget/checkbox.lua:
```diff
@@ -0,0 +1,253 @@
+end
+
+local function draw(self, context, cr, width, height)
+    local size = math.min(width, height)
```
Could `draw` "really use" the provided width/height instead of using the minimum of them? If some layout ignores the fit result and gives this widget a non-rectangular area, some part would otherwise be left "empty" and this could be confusing to users.

Elv13 replied:
No, the problem is when the background shape and check shape isn't the same, it creates a mess as different shapes resize themselves differently. Being a square force some kind of linearity, so the result is predictable(-ish).

psychon replied:
I'd call that a bug in the shape (assuming they also use math.min then the shape should be fixed not to do that), but ok...

psychon commented on tests/examples/wibox/widget/checkbox/custom.lua:
```diff
@@ -0,0 +1,35 @@
+            cr:line_to( rs , rs )
+            cr:move_to( 0  , rs )
+            cr:line_to( rs , 0  )
+            cr:close_path()
```
Why this `close_path`? I'd leave it out. It doesn't really do much here and seems weird to me. The other subpaths aren't explicitly closed either.

psychon commented on lib/gears/shape.lua:
```diff
@@ -376,6 +376,37 @@ function module.losange(cr, width, height)
+    start_angle, end_angle = start_angle or 0, end_angle or math.pi/2
+
+    -- If the shape is a circle, then avoid the lines
+    if math.abs(start_angle + end_angle) == 2*math.pi then
```
How about `>=` instead of `==`?

psychon commented on lib/gears/shape.lua:
```diff
@@ -376,6 +376,37 @@ function module.losange(cr, width, height)
+        cr:line_to(
+            width/2 + math.cos(start_angle)*radius,
+            height/2 + math.sin(start_angle)*radius
+        )
```
Is this `:line_to` really needed? I'm not 100% sure, but I think `:arc` does this automatically.
Also, you could use `rel_line_to` (or whatever the function is called exactly) to get rid of the `width/2+` and `height/2+`.

psychon commented on lib/wibox/container/arcchart.lua:
```diff
@@ -0,0 +1,317 @@
+    self:emit_signal("widget::layout_changed")
+end
+
+--- Get the number of children element
```
`number of`?

psychon commented on lib/wibox/container/arcchart.lua:
```diff
@@ -0,0 +1,317 @@
+--         self:set_max_value(val)
+--     elseif val < 0 then
+--         self:set_min_value(val)
+--     end
```
Any reason not to remove this code?

psychon commented on lib/wibox/container/arcchart.lua:
```diff
@@ -0,0 +1,317 @@
+
+    local delta = 0 - 0
+
+    self._percent = val/delta
```
...or this whole function if it just does a division by zero. Huh?

psychon commented on lib/wibox/container/arcchart.lua:
```diff
@@ -0,0 +1,317 @@
+-- @property max_value
+
+for _, prop in ipairs {"border_width", "border_color", "paddings", "colors",
+    "rounded_edge", "bg", "thickness", "values", "min_value", "max_value" } do
```
Consistency: Here it is called `min/max_value`, some previous widget called it `minimum/maximum_value` (see the comment that I left there). It would be nice if both used the same name.

psychon commented on lib/wibox/container/arcchart.lua:
```diff
@@ -0,0 +1,317 @@
+
+--- Returns a new arcchart layout. A arcchart layout arccharts a given widget. Use
+-- :set_widget() to set the widget and :set_direction() for the direction.
+-- The default direction is "north" which doesn't change anything.
```
There is no direction on this widget either.

Also, "A arcchart layout archarts a given widget" doesn't really say anything...

psychon commented on lib/wibox/widget/slider.lua:
```diff
@@ -0,0 +1,548 @@
+
+function event.mouse_leave(self)
+    self:emit_signal("widget::redraw_needed") --TODO
+--     capi.root.cursor("fleur")
```
This seems unfinished... The cursor is changed, but never reset? Also, why is the cursor on the root window changed? That seems useless and not related to this widget at all. (And no, right now I have no good idea nor much interest to implement per-widget mouse cursors)

psychon commented on lib/wibox/widget/slider.lua:
```diff
@@ -0,0 +1,548 @@
+    -- Fix the X position when used in a rotate or mirror widget
+
+    -- Remove the translation to bring {x, y} and the rect on the same origin
+    local rem_translate = geo.matrix_to_device*gmatrix.identity
```
Multiplication with the identity matrix is a no-op...?

Elv13 replied:
Not since you remove the matrix `:copy()`, even if it's not the cleanest way to get a new matrix to play with. I could revert the revert that deleted the `:copy()` function.

psychon replied:
Why do you need a new matrix to play with? The idea is that matrixes are immutable (= cannot be modified), so there is no reason to copy them. (And this means that modifying a matrix does not accidentally influence some other code that happens to reference the same matrix)

psychon commented on lib/wibox/widget/slider.lua:
```diff
@@ -0,0 +1,548 @@
+    --HACK Cairo 0,0 is top left, a matrix is bottom left, funny...
+    if util.round(rem_translate:transform_point(0, 100)) ~= 0 then
+        px = -(px - rect.width)
+    end
```
I don't get the math here and it doesn't seem to me like the author really understood it either. A matrix doesn't have a "bottom left"...?!?

psychon commented on lib/wibox/widget/slider.lua:
```diff
@@ -0,0 +1,548 @@
+        horizontal = true
+    },
+    widget = wibox.container.mirror
+}]]
```
...????

psychon replied:
...??????

psychon commented on lib/awful/widget/common.lua:
```diff
@@ -84,7 +84,9 @@ function common.list_update(w, buttons, label, data, objects)
 
-        local text, bg, bg_image, icon = label(o, tb)
+        local text, bg, bg_image, icon, args = label(o, tb)
+        args = args or {}
```
:-(
(At least we now have this thing "meta-extensible"!)

psychon commented on lib/awful/widget/taglist.lua:
```diff
@@ -91,15 +94,53 @@ function taglist.taglist_label(t, args)
+
+        if args.shape_border_color_urgent or theme.taglist_shape_border_color_urgent then
+            shape_border_color = args.shape_border_color_urgent or theme.taglist_shape_border_color_urgent
+        end
```
These new values need to be added to the documentation. You do so for the tasklist, but apparently the taglist is forgotten (this is only how I noticed it).

Elv13 replied:
This is documented in a later commit along with some other undocumented things. Mind if it stays that way?

psychon replied:
Ok, fine with me...

psychon commented on docs/common/widget.ldoc:
```diff
@@ -40,6 +40,12 @@
+-- set of containers and layouts wrapping the widget.
+-- @tparam string signal_name
+-- @param ... Other arguments
+-- @function emit_signal_recursive
```
This part is (still) in the wrong commit. It needs to be in the following commit.

psychon commented on lib/wibox/hierarchy.lua:
```diff
@@ -51,6 +51,15 @@ local function hierarchy_new(redraw_callback, layout_callback, callback_arg)
         layout_callback(result, callback_arg)
     end
+    function result._emit_recursive(_, name, ...)
+        local cur = result
```
Didn't you previously have this code in some other PR? My comments from there still apply (I'd make use of the `self` argument instead of ignoring it).

psychon commented on lib/wibox/drawable.lua:
```diff
@@ -144,6 +144,48 @@ local function do_redraw(self)
+
+            cr:restore()
+        end
+    end
```
I'd prefer to have this wibox-related code in `wibox.lua`. How about adding an after-draw-callback that the wibox code could use?

Then again, what does this actually do and how/why does it work? You take the bounding shape of the wibox and somehow apply it to the drawing after it was done? I do not really see how this can work or does work. Also, this CLEARs some pixels, which means that we now get some black where previously there was no black (unless a compositing manager is running...).

So... how does this work? Should it perhaps only be done when a compositing manager is running?

psychon replied:
Actually... could we move this part (hacks(?) to make a wibox shape look nicer) out of this PR and into its own PR? I feel like I'll have to stare a lot at this code.

psychon commented on awesomeConfig.cmake:
```diff
@@ -340,6 +340,13 @@ foreach(file ${AWESOME_CONFIGURE_FILES})
+# {{{ Generate some aggregated documentation from lua script
+execute_process(
+        COMMAND lua ${SOURCE_DIR}/docs/06-appearance.md.lua
+        ${BUILD_DIR}/docs/06-appearance.md
```
Thanks for not doing this in CMake!

psychon commented on docs/06-appearance.md.lua:
```diff
@@ -0,0 +1,222 @@
+
+    for file in function() return enumerator:next_file() end do
+        local file_name = file:get_attribute_as_string(name_attr)
+        local file_type = file:get_file_type()
```
Please look at `lib/menubar/utils.lua`, function `utils.parse_dir`. You have to explicitly request the file type, since otherwise some filesystems (XFS?) do not provide it. Also, instead of giving a string to `enumerate_children`, I would prefer the approach used in menubar since it catches typos (it still provides a string argument, but the string is not hardcoded but comes from gio).

psychon commented on docs/06-appearance.md.lua:
```diff
@@ -0,0 +1,222 @@
+        if file_type == "REGULAR" and file_name:match(ext or "") then
+            table.insert(ret, path..file_name)
+        elseif file_type == "DIRECTORY" then
+            get_all_files(path..file_name.."/", ext, ret)
```
I don't think it matters much, but portability says you shouldn't just concatenate strings, but do what `utils.parse_dir` does: `local child_path = enumerator:get_child(file):get_path()`

psychon commented on docs/06-appearance.md.lua:
```diff
@@ -0,0 +1,222 @@
+        return "../libraries/".. mod ..".html"
+    end
+
+    return ""
```
Should this `error()` out instead of silently producing broken(?) output? (Same for the previous function)

psychon commented on docs/06-appearance.md.lua:
```diff
@@ -0,0 +1,222 @@
+    }
+end
+
+local all_files = get_all_files("/home/lepagee/dev/awesome/lib/", "lua")
```
....

psychon commented on docs/05-awesomerc.md.lua:
```diff
@@ -0,0 +1,241 @@
+   support for static/compile time checks. There is the `luacheck` utility to
+   help find some categories of errors. Those errors will cause Awesome to
+   "drop" the current call stack and start over. Note that if the config cannot
+   reach the end of the `rc.lua` without errors. It will fallback to the
```
`, i` instead of `. I`

psychon commented on lib/wibox/layout/grid.lua:
```diff
@@ -0,0 +1,484 @@
+        table.insert(self._private.children, w)
+    end
+
+    self:emit_signal("widget::layout_changed")
```
Doesn't this also need `widget::redraw_needed` in case a new row was added and there is a border between widgets that has to be redrawn?

psychon commented on lib/wibox/layout/grid.lua:
```diff
@@ -0,0 +1,484 @@
+
+    self._private.rows[row] = self._private.rows[row] or {}
+
+    self:emit_signal("widget::layout_changed")
```
This function doesn't work, does it? You need to add the widget to the row (and perhaps call check_widget())

psychon commented on lib/wibox/layout/grid.lua:
```diff
@@ -0,0 +1,484 @@
+-- @tparam number index The widget index to remove
+-- @treturn boolean index If the operation is successful
+function grid:remove(index)
+    local TODO_TOTAL = #self:get_children()
```
Why the creative variable name?

psychon commented on lib/wibox/layout/grid.lua:
```diff
@@ -0,0 +1,484 @@
+    self:reset()
+    if #children > 0 then
+        self:add(unpack(children))
+    end
```
How about implementing this function as `self:set_children(self:get_children())`? Seems simpler to me.

psychon commented on lib/wibox/layout/grid.lua:
```diff
@@ -0,0 +1,484 @@
+
+    base.check_widget(widget2)
+
+    self._private.rows[index] = widget2
```
No idea what this function should do (it is not documented), but it seems wrong. `rows[index]` should be a table containing widgets (all the widgets in that row), and not just some widget. Is there a test for this function (that also tests if the grid layout can still be drawn afterwards)?

psychon commented on lib/wibox/layout/grid.lua:
```diff
@@ -0,0 +1,484 @@
+-- @param widget The widget
+-- @treturn boolean If the operation is successful
+function grid:insert(index, widget)
+    local TODO_TOTAL = math.huge --FIXME
```
Huh? Please fix O.o

psychon commented on lib/wibox/layout/grid.lua:
```diff
@@ -0,0 +1,484 @@
+    if not index or index < 1 or index > TODO_TOTAL then return false end
+
+    base.check_widget(widget)
+    table.insert(self._private.rows, index, widget) --FIXME
```
This also looks broken. Also, I lost track of the number of ways to add widgets. I would prefer if this function turned the index into row, cell numbers and then used the existing function for setting a widget based on that.

psychon commented on lib/wibox/layout/grid.lua:
```diff
@@ -0,0 +1,484 @@
+        -- The other widgets will be discarded
+        if used_max_h + max_row[row_idx] > orig_height then
+            break
+        end
```
I don't think that the other widgets will be discarded. Instead, I read the code so that they will be "forced into the available space", won't they?

In other words, I think that this `if` and the `break` can be removed without problems (same for the next for loop; no it is "not really" a problem if `:fit` returns numbers that are larger than the available space, `base.fit_widget` will clean that up and force the result into the available space).

psychon commented on lib/wibox/layout/grid.lua:
```diff
@@ -0,0 +1,484 @@
+function grid:reset()
+    self._private.rows = {}
+    self._private.children = {}
+    self:emit_signal("widget::layout_changed")
```
Also: Redraw needed, due to the borders (if there are any)

psychon commented on lib/wibox/layout/grid.lua:
```diff
@@ -0,0 +1,484 @@
+function grid:set_fill_space(val)
+    if self._private.fill_space ~= val then
+        self._private.fill_space = not not val
+        self:emit_signal("widget::layout_changed")
```
I think that this property also influences the border between widgets, so: `redraw_needed`

psychon commented on lib/wibox/layout/grid.lua:
```diff
@@ -0,0 +1,484 @@
+    end
+end
+
+--- The the border (grid) width.
```
The the

psychon commented on lib/wibox/layout/grid.lua:
```diff
@@ -0,0 +1,484 @@
+
+function grid:set_fill_space(val)
+    if self._private.fill_space ~= val then
+        self._private.fill_space = not not val
```
Shouldn't you do the "not not dance" before the `~=`? Not that it matters much, just saying...
So something like:
```
val = not not val
if self._private.fill_space == val then return end
self._private.fill_space = val
etc
```

psychon commented on lib/wibox/layout/grid.lua:
```diff
@@ -0,0 +1,484 @@
+        self._private.border_width = math.max(0, val)
+
+        -- This count as spacing, so the layout really change
+        self:emit_signal("widget::layout_changed")
```
Redraw needed?

psychon commented on lib/wibox/layout/grid.lua:
```diff
@@ -0,0 +1,484 @@
+    end
+
+    return ret
+end
```
Could you move this function to the end of the file?

psychon commented on lib/wibox/layout/grid.lua:
```diff
@@ -0,0 +1,484 @@
+    self:emit_signal("widget::redraw_needed")
+end
+
+local function get_layout(dir, widget1, ...)
```
The `dir` argument should go away, this will be the first widget.

Elv13 replied:
I guess you saw what I used for base at this point. I should cleanup my template.

psychon commented on lib/wibox/container/align.lua:
```diff
@@ -0,0 +1,158 @@
+
+-- Layout this layout
+function align:layout(context, width, height)
+    if not self._private.widget or not self._private.widget._private.visible then
```
I don't think you should check `visible` here. One reason is because it won't make a difference on the visual result. The second reason is "I don't think things will be properly redrawn if the child widget is made visible" (because the widget code doesn't know that this code tries to display the widget). The third reason is "because you are using other code's `_private`".

psychon commented on lib/wibox/container/align.lua:
```diff
@@ -0,0 +1,158 @@
+
+--- Reset this layout. The widget will be removed and the rotation reset.
+function align:reset()
+    self._private.direction = nil
```
There is no direction, is there? Instead this should reset valign and halign.

psychon commented on lib/wibox/container/align.lua:
```diff
@@ -0,0 +1,158 @@
+--- Get the direction of this rotating layout
+function align:get_direction()
+    return self._private.direction or "north"
+end
```
Remove this whole function

psychon commented on lib/wibox/container/align.lua:
```diff
@@ -0,0 +1,158 @@
+    return ret
+end
+
+function align.mt:__call(...)
```
I think this needs `_, ` as a first argument

psychon commented on lib/wibox/widget/graph.lua:
```diff
@@ -142,19 +168,43 @@ function graph.draw(_graph, _, cr, width, height)
+                if step_shape then
+                    local mat = cr:get_matrix()
+                    mat.x0 = x0
+                    cr:set_matrix(mat)
```
Uhm. Could you somehow use `:save()` / `:restore()` instead? For one thing: Because your code wont work inside of a rotating widget. For another thing: You aren't undoing *all*  the translations, but only the translation in x direction.

psychon commented on lib/wibox/widget/graph.lua:
```diff
@@ -142,19 +168,43 @@ function graph.draw(_graph, _, cr, width, height)
+                    if step_shape then
+                        cr:translate(step_width + (i>1 and step_spacing or 0), height * (1 - value))
+                        step_shape(cr, step_width, height)
+                        cr:translate(nil, -(height * (1 - value)))
```
I *bet*  that lgi / cairo do not let you call this function with a `nil` argument. This should likely be 0 and needs some test which catches this breakage.

Elv13 replied:
bet taken :P but I will change it anyway.

psychon replied:
Wow, you are right. That's a surprise. I always saw lgi as being quite strict...

psychon commented on lib/wibox/container/align.lua:
```diff
@@ -0,0 +1,158 @@
+---------------------------------------------------------------------------
+-- A container used to align smaller widgets into larger space.
```
This gives us a container called align and a layout called align, making the name "align" ambiguous. This needs another name (but I don't have any good suggestions, sorry).

psychon commented on lib/naughty/core.lua:
```diff
@@ -107,14 +107,20 @@ naughty.config.presets = {
 -- @tfield[opt="top_right"] string position
+-- @tfield[opt] gears.color border_color
+-- @tparam[opt] gears.shape shape Widget shape.
+-- @tparam[opt] table shape_args Arguments passed to widget shape function.
```
Why this mix of tfield and tparam? I think all of this should be tfield...?

psychon commented on awesomeConfig.cmake:
```diff
@@ -347,7 +347,7 @@ execute_process(
 execute_process(
         COMMAND lua ${SOURCE_DIR}/docs/05-awesomerc.md.lua
-        ${BUILD_DIR}/05-awesomerc.md.lua ${SOURCE_DIR}/awesomerc.lua
+        ${BUILD_DIR}/docs/05-awesomerc.md ${SOURCE_DIR}/awesomerc.lua
```
This looks like a change/fix that should be squashed into the original commit instead. Same for the other changes in this commit?

psychon commented on docs/06-appearance.md.lua:
```diff
@@ -10,7 +10,7 @@ local name_attr = gio.FILE_ATTRIBUTE_STANDARD_NAME
 local function get_all_files(path, ext, ret)
     ret = ret or {}
-    local enumerator = gio.File.new_for_path(path):enumerate_children(
+    local enumerator, _ = gio.File.new_for_path(path):enumerate_children(
```
...why?

psychon commented:
I read through this up to 2b3c2d746979c86a2d23e46ce12f426224a0ccbe (just so that I know which version of this I looked at). I know that this is still titled WIP, but I took a look anyway.

Elv13 commented:
Ok, I addressed 5-6 of @psychon comments. I fixed all luachecks, made it work with lua5.3 and improved the `grid` layout to fix some more broken corners cases (and wrote the tests for them). This is a longterm-ish PR, so don't expect all issues to be fixed by tomorrow. However, it is progressing well. And the 93% coverage is getting nice too! For some reason Codecov is missing the integration test coverage (`test-wibox.lua`). By the "old" way codecov worked before being broken(?) with a commit yesterday, I did hit 75% (and yes, it is because it include the tests themselves in the percentage, so adding 1.5k lines of tests did artificially boost the perceived coverage).

Elv13 commented:
Ok, I addressed most issues above (I think).

The slider need to work I guess. I think I will remove from the PR is @psychon think it still suck too much. 

@actionless Do you mind if your patches go into a new PR instead of this one along with the rounded corner one? You can open that pull request yourself so you can work on it.

I removed the [WIP] tag. I guess it will need another row of review, but I think it is close enough to be merged. About the "new" 1.1K lines since the last update, they are all tests or doc.

psychon commented:
One of the travis build segfaulted in `cairo_region_union_rectangle`. Scary, but I can't see how this PR could cause that. Hm... :-(

> The slider need to work I guess. I think I will remove from the PR is @psychon think it still suck too much. 

Without having looked at the slider widget:
This PR is titled "shape api" and adds a slider widget.

> I guess it will need another row of review,

I'll try to find some time to go through this on Sunday, but I cannot promise yet.

actionless commented:
@Elv13 yup, i don't mind, anyway i am using my fork for building the local package instead of upstream repo

Elv13 commented:
> This PR is titled "shape api" and adds a slider widget.

The PR add widgets using the shape API and retrofit existing ones to support it. Beside all the documentation stuff (that was kind of necessary to be able to properly document those widgets), the only unrelated code is the grid layout (as said in the first message).

psychon commented on lib/wibox/container/radialprogressbar.lua:
```diff
@@ -0,0 +1,264 @@
+    cr:translate(wa.x, wa.y)
+    shape.rounded_bar(cr, wa.width, wa.height)
+    cr:clip()
+    cr:translate(-wa.x, -wa.y)
```
I don't think this works correctly. The `wibox.hierarchy` now has a wrong idea on where the child widget is and so partial redraws will not redraw the correct part of the screen.

However, your `:layout()` already applies a translation (in fact, the opposite translation) to the child widget, so why is this needed at all?

psychon commented on lib/wibox/container/radialprogressbar.lua:
```diff
@@ -0,0 +1,264 @@
+function radialprogressbar:fit(context, width, height)
+    if self._private.widget then
+        local wa = content_workarea(self, width, height)
+        return base.fit_widget(self, context, self._private.widget, wa.width, wa.height)
```
I'm not totally sure, but I think this should be:
```lua
local w, h = base.fit_widget(self, context, self._private.widget, wa.width, wa.height)
return wa.x + w, wa.y + h
```

psychon commented on lib/wibox/container/radialprogressbar.lua:
```diff
@@ -0,0 +1,264 @@
+    radialprogressbar["set_"..v.."_padding"] = function(self, val)
+        self._private.paddings = self._private.paddings or {}
+        self._private.paddings[v] = val
+        self:emit_signal("widget::redraw_needed")
```
Also `widget::layout_changed`, because the `content_workarea()` depends on the padding and is used in `:layout()`

psychon commented on lib/wibox/container/radialprogressbar.lua:
```diff
@@ -0,0 +1,264 @@
+    return ret
+end
+
+function radialprogressbar.mt:__call(...)
```
This should be `_, ...` instead of `...` since the first argument is just `wibox.container.radialprogressbar`

psychon commented on lib/wibox/widget/checkbox.lua:
```diff
@@ -0,0 +1,252 @@
+end
+
+local function draw(self, _, cr, width, height)
+    local size = math.min(width, height)
```
This isn't needed (at least in `draw`), because this only ever uses `size, size` which can easily be replaced with `width, height`. Only `fit` needs this. And yes, I know that you disagree.

psychon commented on lib/wibox/widget/piechart.lua:
```diff
@@ -0,0 +1,212 @@
+
+local function draw_label(cr,angle,radius,center_x,center_y,text)
+    cr:move_to(center_x+(radius/2)*math.cos(angle),center_y+(radius/2)*math.sin(angle))
+    cr:line_to(center_x+(1.5*radius)*math.cos(angle),center_y+(1.5*radius)*math.sin(angle))
```
Perhaps `rel_line_to` helps here? The above `line_to` can be replaced with (at least I think so):
```lua
cr:rel_line_to(radius*math.cos(angle), radius*math.sin(angle))
```

psychon commented on lib/wibox/widget/piechart.lua:
```diff
@@ -0,0 +1,212 @@
+    cr:move_to(center_x+(radius/2)*math.cos(angle),center_y+(radius/2)*math.sin(angle))
+    cr:line_to(center_x+(1.5*radius)*math.cos(angle),center_y+(1.5*radius)*math.sin(angle))
+    local x,y = cr:get_current_point()
+    cr:line_to(x+(x>center_x and radius/2 or -radius/2),y)
```
```lua
cr:rel_line_to(x > center_x and radius/2 or -radius/2, 0)
```

psychon commented on lib/wibox/widget/piechart.lua:
```diff
@@ -0,0 +1,212 @@
+        extents = cr:font_extents()
+    else
+        extents = cr.font_extents
+    end
```
I would suggest:
```lua
local extents = cr.font_extents
-- Support older LGI versions
if type(extents) == "function" then
  extents = extents(cr)
end

psychon commented on lib/wibox/widget/piechart.lua:
```diff
@@ -0,0 +1,212 @@
+        extents = cr.font_extents
+    end
+
+    cr:move_to(x+(x>center_x and radius/2 + 5 or (-radius/2 - cr:text_extents(text).width - 5)),y+(extents.height/4))
```
I'd like to have more explanation for this. Why 5? (likely just some distance between the text and the line, perhaps replacing this with a space character in the text?) Why `extents.height/4`? This feels like some magic numbers that hide some incorrectness.

psychon commented on lib/wibox/widget/piechart.lua:
```diff
@@ -0,0 +1,212 @@
+--@DOC_object_COMMON@
+
+return setmetatable(module, { __call = function(_, ...) return new(...) end })
+-- kate: space-indent on; indent-width 4; replace-tabs on;
```
Nice try, but I saw it :-)

psychon commented on lib/gears/shape.lua:
```diff
@@ -331,6 +331,7 @@ end
 -- @tparam[opt=math.min(width  height) / 2)] number radius The radius
 function module.circle(cr, width, height, radius)
     radius = radius or math.min(width, height) / 2
+    cr:move_to(width/2+radius, height/2)
```
What happens without this? Looking at cairo's implementation (`_cairo_default_context_arc()`), the first thing that it does is a `line_to` to the start point of the arc. If you just want to get rid of this `line_to`, I would suggest `cr:new_sub_path()` instead (easier to read/understand).

psychon commented on lib/wibox/container/arcchart.lua:
```diff
@@ -0,0 +1,343 @@
+    arcchart["set_"..v.."_padding"] = function(self, val)
+        self._private.paddings = self._private.paddings or {}
+        self._private.paddings[v] = val
+        self:emit_signal("widget::redraw_needed")
```
Doesn't this also need `layout_changed`?

psychon commented on lib/wibox/widget/slider.lua:
```diff
@@ -0,0 +1,528 @@
+
+-- Move the handle to the correct location
+local function move_handle(self, x, y, geo)
+    -- Fix the X position when used in a rotate or mirror widget
```
Uhm, I'll take this as a bug report. x and y should already be in the widget's own space, but I see that the code in `drawable` for this is wrong (completely ignoring rotations). Do you want to fix this or should I do? Something like the following might work (untested):
```lua
diff --git a/lib/wibox/drawable.lua b/lib/wibox/drawable.lua
index 0212efa..6b450bb 100644
--- a/lib/wibox/drawable.lua
+++ b/lib/wibox/drawable.lua
@@ -376,8 +376,7 @@ function drawable.new(d, widget_context_skeleton, drawable_name)
             local widgets = ret:find_widgets(x, y)
             for _, v in pairs(widgets) do
                 -- Calculate x/y inside of the widget
-                local lx = x - v.x
-                local ly = y - v.y
+                local lx, ly = v.matrix_to_device:invert():transform_point(x, y)
                 v.widget:emit_signal(name, lx, ly, button, modifiers,v)
             end
         end)
```
(find_widgets should just be taught to also return the matrix_from_device so that this `invert()`-call isn't needed, but whatever...)

psychon commented:
I read up to f4b32faac45b268d998ccd149e2a68a3dc4157e5 and now ran out of review-time. Sorry, but could this be split up into at least one other PR? For example, the new shapes and widgets seem relatively independent from the docs-related changes. Simplest way would be to just open a separate PR for half of the commits here so that this doesn't cause much work for you.
Thanks & sorry.

actionless commented:
i am trying to get this (https://github.com/awesomeWM/awesome/pull/1044/commits/0ef77e268fc130b26df93263f53b142fe877af83) working on master branch again, could you hint me which parts of this changeset are currently missing for that?

Elv13 commented:
The client shape commit, just about everything else is merged

actionless commented:
weird, when i am re-applying naughty-shape commit on top of master i don't see any warnings but notification shape stays rectangular

also in documentation wibox do not have `shape_border_width`, `shape` and `shape_args`: http://new.awesomewm.org/apidoc/classes/wibox.html

Elv13 commented:
Let me rephrase myself. Almost have commits have been merged *at the exception* of the mentioned wibox shape commit. It has not been merged because it needs work. Work I would like you to do if you wish to get your notification commit merged. It isn't much, but it is delicate and I already spent enough time getting the thing to "work".

actionless commented:
that's why i was asking you to hint me from where to start :)

Elv13 commented:
Look at the comments above (in the PR) related to this commit, that's
pretty much it ;)

On 8 October 2016 at 06:18, Yauhen Kirylau <notifications@github.com> wrote:

> that's why i was asking you to hint me from where to start :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/awesomeWM/awesome/pull/1044#issuecomment-252416808>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAUxoG86w-31gG9A4M1gxXk4ErmxlOC2ks5qx24LgaJpZM4JlE9A>
> .
>

Elv13 commented:
this PR was finally fully merged in 4.1
